### PR TITLE
Micromega clean-up

### PR DESCRIPTION
**Pull request by maximedenes**
We add .mli files, removed dead code and use standard combinators
instead of redefined ad-hoc ones in a few places.

A lot of cleaning still has to be done on this code: documenting the
interfaces, resolving the many abstraction leaks. I suspect there is
still a lot of code duplication.

@Zimmi48 I'm not sure which "part" label I should use.

### CODE REVIEW

**Comment by maximedenes** [comment]
@Zimmi48 I see that @fajb was not automatically asked a review as a code owner. Does it mean we should add him to the Coq organization?

**Comment by Zimmi48** [comment]
@fajb never replied to the question whether he agreed to be a code owner in #7032. Unless you have more direct ways to reach him, we may need to find another principal maintainer for micromega.

**Comment by maximedenes** [comment]
Did you drop him an e-mail?

**Comment by Zimmi48** [comment]
No, I didn't. I only did it for the only person who didn't have a GitHub account already.

On your previous question (which "part" label), the answer is: it's always OK to create new labels.

**Comment by fajb** [comment]
Hello, 

I have received the previous notifications...
For the "code owner" thing, I received the mail during vacations and, since then,  did not find time to read the instructions...
Sure, I'd rather like to keep an eye on the evolution of the code and be notified if it something happens.
If there is a bug report, I can also spend some time to fix it.
Yet, from my side, I do not intend to improve it. This is unfortunate as, I agree, there would be much cleaning to do.
(But also much to do to get eventually a decent efficient and complete decision procedure.)

Anyway, I had a quick look at your modifications. They are fine by me except, maybe, that printing code seems to have disappeared. Would it be useful for debugging purposes? Or is-it simply redundant with code existing elsewhere ?
Anyway, thanks for the maintenance.

Best,
PS: If you look at my commit history, it seems I could live without a commit right that I may only use once every few years.

**Comment by maximedenes** [comment]
> They are fine by me except, maybe, that printing code seems to have disappeared. Would it be useful for debugging purposes?

Thanks for the feedback, @fajb ! A lot of the debugging printing code is already plugged in, under an `if debug` conditional. I didn't remove that part, only what was really dead code. Maybe you could tell me what printing functions look useful to you, we should bring them back and probably register them as printers for `ocamldebug` (I don't think they were).

**Comment by fajb** [comment]
Keep it as it is.
If they were useful, time will tell!

**Comment by Zimmi48** [comment]
Hello @fajb! Nice to know that you are correctly receiving the notifications. https://github.com/coq/coq/blob/master/dev/doc/MERGING.md shouldn't be too long to read but one important point is that this code owner / maintainer status does not require you to actively commit code, only to review and merge other people's changes. If you think that's doable, that would be great! Otherwise, your reviews are still very much appreciated.

**Comment by maximedenes** [comment]
@SkySkimmer I'm pretty sure I ran `git rebase --whitespace=fix`. Isn't it supposed to fix all whitespace errors the linter can find?

**Comment by Zimmi48** [comment]
Not the blank line problems at EOF which is what is reported here:

```
plugins/micromega/sos.ml:1112: new blank line at EOF.
```

**Comment by maximedenes** [comment]
Ok, this is a dealbreaker for me. I don't want to do manual whitespace fixes, especially that I don't see any issue with one blank line. Will report an issue separately.

**Comment by maximedenes** [comment]
Done in #7463.

**Comment by maximedenes** [comment]
@Zimmi48 This PR is a bit special, as it was "morally" approved by @fajb, but we don't explicitly have a principal maintainer. Can you merge it? If it passes CI, of course.

**Comment by Zimmi48** [comment]
I can. Although if @fajb doesn't want to assume the role of principal maintainer, we have to find someone else.

**Comment by Zimmi48** [comment]
Can you rebase to have a recent CI build?

**Comment by maximedenes** [comment]
> Can you rebase to have a recent CI build?

I rebased today. Isn't it recent enough?

> I can. Although if @fajb doesn't want to assume the role of principal maintainer, we have to find someone else.

Indeed, we should clarify that. But meanwhile, I'd prefer to see this non-invasive PR merged. Otherwise I'll close it, because I need to reduce my todo-list.

**Comment by Zimmi48** [comment]
That's weird. It looks like the linter fix in #7464 did not actually fix the problem it was supposed to fix.
cc @SkySkimmer

**Comment by SkySkimmer** [comment]
>That's weird. It looks like the linter fix in #7464 did not actually fix the problem it was supposed to fix.

Seems like somehow `blank-at-eol` is what complains about "new blank line at EOF".

**Comment by Zimmi48** [comment]
Then it's going to complain also on Travis branch build after we merge, and in every subsequent linter test, right? In this case, I cannot merge as-is. `git` is not very helpful here: have a newline at EOF is actually helpful so we do not want to disable this test completely, but it is testing more and something that it can't fix automatically :disappointed:

**Comment by SkySkimmer** [comment]
No this is pre-commit.

 >we do not want to disable this test completely

We already did in #7464 though.

Gaëtan Gilbert

**Comment by Zimmi48** [comment]
Sorry, I didn't understand your reply.

**Comment by SkySkimmer** [comment]
The part of the linter which errors is run on the diff for each commit in a PR, so future PRs won't be impacted.

>having a newline at EOF is actually helpful so we do not want to disable this test completely

>We already did in #7464 though.

My mistake, that test is not disabled and not related to the error, it's implemented by our custom script and not through git.

**Comment by Zimmi48** [comment]
OK thanks for the info. I'll merge the PR now then.

Regarding why the linter fix didn't work, this is very strange. Git's documentation says "trailing-space being a short-hand to cover both blank-at-eol and blank-at-eof" which seems to confirm that your fix at #7464 was the right one. If we have this problem again in the future, we should investigate if it's not a bug in git.

File: plugins/micromega/csdpcert.mli [hunk 1]
```diff
@@ -0,0 +1,9 @@
+(************************************************************************)
+(*         *   The Coq Proof Assistant / The Coq Development Team       *)
+(*  v      *   INRIA, CNRS and contributors - Copyright 1999-2018       *)
+(* <O___,, *       (see CREDITS file for the list of authors)           *)
+(*   \VV/  **************************************************************)
+(*    //   *    This file is distributed under the terms of the         *)
+(*         *     GNU Lesser General Public License Version 2.1          *)
+(*         *     (see LICENSE file for the text of the license)         *)
+(************************************************************************)
```
Comment (Zimmi48) [inline plugins/micromega/csdpcert.mli]: What does it mean when a .ml exports nothing at all? Doesn't it mean that the .ml file could be removed completely?

Comment (SkySkimmer) [reply plugins/micromega/csdpcert.mli]: It's only there for the side effects.

Comment (SkySkimmer) [reply plugins/micromega/csdpcert.mli]: Or rather this file is compiled into a standalone executable.

Comment (Zimmi48) [reply plugins/micromega/csdpcert.mli]: OK thanks.

**Pull request merged**